cfripper/cli.py

```diff
@@ -0,0 +1,125 @@
+    cfmodel = pycfmodel.parse(template)
+    return cfmodel
+
+
+def process_template(
+    template, resolve: bool, resolve_parameters: Optional[Dict], output_folder: Optional[str], output_format: str
+) -> None:
+    logging.info(f"Analysing {template.name}...")
+    cfmodel = get_cfmodel(template)
+    if resolve:
+        cfmodel = cfmodel.resolve(resolve_parameters)
+    config = Config(rules=DEFAULT_RULES.keys())
+    result = Result()
+    rules = [DEFAULT_RULES.get(rule)(config, result) for rule in config.rules]
+    rule_processor = RuleProcessor(*rules)
+
+    rule_processor.process_cf_template(cfmodel, config, result)
+
+    if output_format == "json":
+        formatted_result = result.json()
+    else:
+        formatted_result = f"Valid: {result.valid}"
+        if not result.valid:
+            formatted_result += "\n" + "\n".join(fr.reason for fr in result.failed_rules)
+
+    if output_folder:
+        output_file = Path(output_folder) / f"{template.name}.cfripper.results.{output_format}"
+        output_file.write_text(formatted_result)
+        logging.info(f"Result saved in {output_file}")
+    else:
+        click.echo(formatted_result)
+
+
+@click.command()
+@click.version_option(prog_name="cfripper", version=__version__)
+@click.argument("templates", type=click.File("r"), nargs=-1)
+@click.option(
+    "--resolve/--no-resolve",
+    is_flag=True,
+    default=False,
+    help="Resolves cloudformation intrinsic functions",
+    show_default=True,
+)
+@click.option(
+    "--resolve-parameters",
+    type=click.File("r"),
+    help=(
+        "JSON/YML file containing key-value pairs used for resolving CloudFormation files with templated parameters. "
+        'For example, {"abc": "ABC"} will change all occurrences of {"Ref": "abc"} in the CloudFormation file to "ABC".'
+    ),
+)
+@click.option(
+    "--format",
+    "output_format",
+    type=click.Choice(["json", "txt"], case_sensitive=False),
+    default="txt",
+    help="Output format",
+    show_default=True,
+)
+@click.option(
+    "--output-folder",
+    type=click.Path(exists=True, resolve_path=True, writable=True, file_okay=False),
+    help="If not present, result will be sent to stdout",
+)
+@click.option(
+    "--logging",
+    "logging_level",
+    type=click.Choice(LOGGING_LEVELS.keys(), case_sensitive=True),
+    default="INFO",
+    help="Logging level",
+    show_default=True,
+)
+def cli(templates, logging_level, resolve_parameters, **kwargs):
+    """Analyse AWS Cloudformation templates passed by parameter."""
+    try:
+        setup_logging(logging_level)
+
+        if kwargs["resolve"] and resolve_parameters:
+            resolve_parameters = convert_json_or_yaml_to_dict(resolve_parameters.read())
+
+        for template in templates:
+            process_template(template=template, resolve_parameters=resolve_parameters, **kwargs)
+
+    except Exception as e:
+        logging.error(str(e))
+        logging.debug("", exc_info=True)
```

Comment:
Why don't you just use logging.exception? Having to rerun as debug to get some usable output feels like a waste

Reply:
I was thinking to separate stack trace and error title. If a user runs the CLI in `ERROR` mode doesn't want to have the debug in the logs (in this case the stack trace).
I can understand that if an exception was unhandled and breaks the CLI to report always